File: python/http/invoker.py
"""Enable a function to respond to HTTP requests.

Converts a function into a Flask app. Incoming requests on the specified port
will cause the function to be invoked.

TODO:
- read module and entrypoint from environment variables
- move invoker.py (and Dockerfile/gunicorn.conf?) to python/
"""


from importlib import import_module
from flask import Flask, request, make_response


FUNCTION_MODULE = 'main'
FUNCTION_ENTRYPOINT = 'hello_http'

def load_user_function(module, entrypoint):
    """Load the user's function and return it to the caller."""
    mod = import_module(module)
    user_func = getattr(mod, entrypoint)
    return user_func


# Load the user's function
user_function = load_user_function(FUNCTION_MODULE, FUNCTION_ENTRYPOINT)

# Create a Flask app that invokes the user's function in response to
# requests on path '/'
app = Flask(__name__)


@app.route('/')
def invoke_user_function():
    """Invoke the user's function."""
    response = user_function(request)
    if not response:
        response = ''
    return make_response(response)

File: nodejs6/http/invoker.js
FUNCTION_MODULE = 'index.js'
FUNCTION_ENTRYPOINT = 'helloHttp'

function loadUserFunction(module, entrypoint) {
    const userFunction = require(`./${module}`)[entrypoint];
    return userFunction
}

// Load the user's function
const userFunction = loadUserFunction(FUNCTION_MODULE, FUNCTION_ENTRYPOINT);

// Create an Express app
const express = require('express');
const app = express();
const bodyParser = require('body-parser');

app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: false }));

// Invoke the user's function on any request
app.all('*', userFunction);

const host = '0.0.0.0'
const port = '8000'
const server = app.listen(port, () => {
    console.log(`Listening at http://${host}:${port}`);
});